### PR TITLE
Added project_definition support

### PR DESCRIPTION
**Pull request by danpawlik**
This commit provides project_definition support. It means
that you are able to define project in your index definition,
and monocle will return only informations related to
repositories-regex [1]. For example:

tenants:
  - index: openstack
    crawler:
      loop_delay: 600
      gerrit_repositories:
        - name: "^openstack/.*"
          updated_since: "2021-01-01"
          base_url: https://review.opendev.org/
    projects:
      - name: infra
        repositories-regex: "opendev/system-config|openstack/disskimage-builder"

The infra project definition returns statistics related only for those
repositories that are in repositories-regex.

[1] https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-regexp-query.html

### CODE REVIEW

File: monocle/webapp.py [hunk 1]
```diff
@@ -121,11 +121,37 @@ def whoami():
...
-@app.route("/api/0/query/<name>", methods=["GET"])
-def query(name):
+def _get_index(request):
     if not request.args.get("index"):
         abort(make_response(jsonify(errors=["No index provided"]), 404))
-    index = request.args.get("index")
+    return request.args.get("index")
+
+
+def _get_project_definition(index):
```
Comment (morucci) [inline monocle/webapp.py:130]: You should read the ProjectConfig the same way than in line 81. And put the data in global (yes that's bad, there is a single place on that file where we read the configuration file so better to not duplicate).

File: monocle/webapp.py [hunk 2]
```diff
@@ -121,11 +121,37 @@ def whoami():
...
+    return (
+        config.get_project_definition(monocle_config, index) if monocle_config else None
+    )
+
+
+@app.route("/api/0/config_project_def", methods=["GET"])
+def get_cfg_project_definition():
+    index = _get_index(request)
+    projects = _get_project_definition(index)
+    if not projects:
```
Comment (morucci) [inline monocle/webapp.py:141]: If no project definition then simply return an empty list

File: monocle/webapp.py [hunk 3]
```diff
@@ -121,11 +121,37 @@ def whoami():
...
+
+
+def _get_project_definition(index):
+    monocle_config = yaml.safe_load(open(config_path)) if config_path else None
+    return (
+        config.get_project_definition(monocle_config, index) if monocle_config else None
+    )
+
+
+@app.route("/api/0/config_project_def", methods=["GET"])
```
Comment (morucci) [inline monocle/webapp.py:137]: I prefer such simply endpoint name.
@app.route("/api/0/projects", ...

Comment (morucci) [reply monocle/webapp.py:137]: Also this need to be protected by the ACL

File: monocle/webapp.py [hunk 4]
```diff
@@ -152,8 +178,17 @@ def query(name):
...
     params = utils.set_params(args)
+
+    if params["project_definition"] and not projects:
+        return (
+            "There are no project definion set in "
+            "config file: %s" % request.get("index"),
+            404,
+        )
+
+    params["config_project_definitions"] = projects
```
Comment (morucci) [inline monocle/webapp.py:191]: Not sure why we need that. We already have the endpoint that is api/0/projects so that's enough.

File: monocle/webapp.py [hunk 5]
```diff
@@ -121,11 +121,37 @@ def whoami():
...
+            404,
+        )
+    return jsonify(projects)
+
+
+@app.route("/api/0/query/<name>", methods=["GET"])
+def query(name):
+
+    index = _get_index(request)
+    projects = _get_project_definition(index)
```
Comment (morucci) [inline monocle/webapp.py:153]: Not needed.

File: monocle/utils.py [hunk 6]
```diff
@@ -208,6 +208,7 @@ def getter(attr, default):
     params["has_issue_tracker_links"] = getter("has_issue_tracker_links", None)
     params["change_ids"] = getter("change_ids", None)
     params["target_branch"] = getter("target_branch", None)
+    params["project_definition"] = getter("project_definition", None)
```
Comment (morucci) [inline monocle/utils.py:211]: Not needed.

File: monocle/db/queries.py [hunk 7]
```diff
@@ -72,6 +72,7 @@
     "new_contributors",
```
Comment (morucci) [inline monocle/db/queries.py:72]: I don't see the need to change something in that file.

File: monocle/tests/unit/test_api.py [hunk 8]
```diff
@@ -140,3 +142,86 @@ def test_whoami(self):
             sess["username"] = "jane"
         resp = self.client.get("/api/0/whoami")
         self.assertEqual("jane", json.loads(resp.data))
+
```
Comment (morucci) [inline monocle/tests/unit/test_api.py:145]: Only the api/0/projects endpoint must be validated.

**Review by morucci — CHANGES_REQUESTED** [review]
I think we only need the config schema + the api/0/projects endpoint + the unittests for the endpoint. I don't get the changes in the db/queries.py.

The UI will get a component that query this new endpoint to get projects definitions for a given index.

File: monocle/webapp.py (hunk 1 above)
Comment (danpawlik) [reply monocle/webapp.py:130]: I really don't like. Putting something as a global is going to raise some problems, but as you prefer.

File: monocle/webapp.py (hunk 2 above)
Comment (danpawlik) [reply monocle/webapp.py:141]: Then in react will complicate next pull request, where axios will get array with promise object that it will comes to undefined and it will other issues. Can not be like it is now?

File: monocle/config.py [hunk 9]
```diff
@@ -204,6 +219,10 @@ class Username(str):
     pass
 
 
+class ProjectDefinition(str):
```
Comment (morucci) [inline monocle/config.py:222]: @dataclass
class ProjectDefinition:
    name: str
    repositories_regex: str

File: monocle/db/queries.py [hunk 10]
```diff
@@ -1023,3 +1024,21 @@ def repos_summary(es, index, repository_fullname, params):
         )
         repos[name]["changes_merged"] = count_merged_changes(es, index, name, params)
     return {"summary": repos}
+
+
```
Comment (morucci) [inline monocle/db/queries.py:1028]: Not needed

File: monocle/webapp.py (hunk 2 above)
Comment (morucci) [reply monocle/webapp.py:139]: ok

File: monocle/webapp.py [hunk 11]
```diff
@@ -154,6 +177,16 @@ def query(name):
...
+    projects = project_defs[index] if index in project_defs else []
+
+    if params["project_definition"] and not projects:
+        return (
+            "There are no project definion set in "
+            "config file for : %s" % params["project_definition"],
+            404,
+        )
+
+    params["config_project_definitions"] = projects
```
Comment (morucci) [inline monocle/webapp.py:189]: Not needed

**Review by morucci — CHANGES_REQUESTED** [review]

File: monocle/webapp.py [hunk 12]
```diff
@@ -154,6 +177,19 @@ def query(name):
...
+        return (
+            "There are no project definion set in "
+            "config file for : %s" % params["project_definition"],
+            404,
+        )
+
+    for project_def in projects:
+        if project_def.get("name") == params.get("project_definition"):
+            repository_fullname = project_def.get("repositories_regex")
+
```
Comment (morucci) [inline monocle/webapp.py:192]: Not needed.

**Review by morucci — CHANGES_REQUESTED** [review]
We don't need a new query parameter "project_definition". All parameters to filter based on a project definition already exists.

File: monocle/tests/unit/test_api.py [hunk 13]
```diff
@@ -140,3 +142,88 @@ def test_whoami(self):
...
+            )
+            resp_data = json.loads(resp.data)
+            self.assertEqual(1, resp_data["total"])
+
+            # set wrong index
+            fake_index = "some-fake-index"
+            resp = self.client.get(
+                "/api/0/query/changes?index=%s&project_definition=test" % fake_index
+            )
+            self.assertEqual(404, resp.status_code)
```
Comment (morucci) [inline monocle/tests/unit/test_api.py:189]: This test is not needed.

**Review by morucci — CHANGES_REQUESTED** [review]
See inline comments.

File: README.md [hunk 14]
```diff
@@ -378,6 +378,16 @@ To check that, you are able to run simple curl command:
...
+## Using API
+
+Below there are describe few examples, how to use API in Monocle:
+
+```
+curl -XGET "http://localhost:9876/api/0/query/last_changes?index=zuul&repository=.*&gte=2020-12-24&from=0&size=100&state=OPEN&selfMerged=false&ec_same_date=true"
+curl -XGET "http://localhost:9876/api/0/projects?index=openstack"
+curl -XGET "http://localhost:9876/api/0/query/changes?index=monocle&project_definition=test"
+```
+
```
Comment (morucci) [inline README.md:390]: I think this should be handled in another PR dedicated to improve the documentation for the API usage.

**Review by morucci — CHANGES_REQUESTED** [review]
That's look better IHMO now :) Just few things:
- Could you move the documentation of the API in another PR. This will be a place holder to have a more complete API definition doc in the README.md ?
- Could you add a Changelog entry ?

Then I'll merge the change.

**Review by morucci — APPROVED** [review]
